src/biogeophys/HillslopeHydrologyMod.F90

```diff
@@ -0,0 +1,1140 @@
+    use spmdMod         , only : masterproc
+    use clm_varcon      , only : spval, ispval, grlnd 
+    use landunit_varcon , only : istsoil
+    use clm_time_manager, only : get_step_size_real
+
+    !
+    ! !ARGUMENTS:
+    type(bounds_type), intent(in) :: bounds
+    type(waterstatebulk_type), intent(inout) :: waterstatebulk_inst
+    type(waterfluxbulk_type),  intent(inout) :: waterfluxbulk_inst
+    type(wateratm2lndbulk_type), intent(inout) :: wateratm2lndbulk_inst
+
+    integer               :: c, l, g, i, j
+    real(r8) :: qflx_surf_vol                           ! volumetric surface runoff (m3/s)
+    real(r8) :: qflx_drain_perched_vol                  ! volumetric perched water table runoff (m3/s)
+    real(r8) :: qflx_rsub_sat_vol                       ! volumetric correction runoff (m3/s)
+    real(r8) :: dtime                                   ! land model time step (sec)
+
+    character(len=*), parameter :: subname = 'HillslopeUpdateStreamWater'
+
+    !-----------------------------------------------------------------------
+    associate(                                                            & 
+         stream_water_volume     =>    waterstatebulk_inst%stream_water_lun    , & ! Input/Output:  [real(r8) (:)   ] stream water volume (m3)
+         qstreamflow             =>    waterfluxbulk_inst%qstreamflow_lun      , & ! Input:  [real(r8) (:)   ] stream water discharge (m3/s)
+         qdischarge              =>    waterfluxbulk_inst%qdischarge_col       , & ! Input: [real(r8) (:)   ]  discharge from columns (m3/s)
+         qflx_drain_perched      =>    waterfluxbulk_inst%qflx_drain_perched_col, &! Input:  [real(r8) (:)   ]  column level sub-surface runoff (mm H2O /s)
+         qflx_rsub_sat           =>    waterfluxbulk_inst%qflx_rsub_sat_col    , & ! Input:  [real(r8) (:)   ]  column level correction runoff (mm H2O /s)
+         qflx_surf               =>    waterfluxbulk_inst%qflx_surf_col        , & ! Input: [real(r8) (:)   ]  total surface runoff (mm H2O /s)
+         stream_water_depth      =>    waterstatebulk_inst%stream_water_depth_lun & ! Output:  [real(r8) (:)   ] stream water depth (m)
+         )
+
+       ! Get time step
+       dtime = get_step_size_real()
+
+       do l = bounds%begl,bounds%endl
+          if(lun%itype(l) == istsoil) then
+             g = lun%gridcell(l)
+             
+             do c = lun%coli(l), lun%colf(l)
+                if (col%is_hillslope_column(c)) then            
+                   qflx_surf_vol = qflx_surf(c)*1.e-3_r8 &
+                        *(grc%area(g)*1.e6_r8*col%wtgcell(c))
+                   qflx_drain_perched_vol = qflx_drain_perched(c)*1.e-3_r8 &
+                        *(grc%area(g)*1.e6_r8*col%wtgcell(c))
+                   ! rsub_sat is not included in qdischarge, so add explicitly
+                   qflx_rsub_sat_vol = qflx_rsub_sat(c)*1.e-3_r8 &
+                        *(grc%area(g)*1.e6_r8*col%wtgcell(c))
```

Comment:
- [ ] The multiplication by col%wtgcell to create volumetric quantities highlights a potential issue here that we should at least discuss and possibly fix: I took pains a number of years ago to ensure that a zero-weight (virtual) vegetated landunit would operate effectively the same as a non-zero-weight landunit. This is needed for the glacier coupling and other aspects of dynamic landunits. Multiplying by col%wtgcell *might* be a problem in this respect. I think it depends on how these terms (stream_water_volume, qstreamflow, stream_water_depth and anything they affect) are used in the code. If they are just used for diagnostic purposes and to send gridcell average quantities to ROF, then I think this is fine. But if they feed back onto the column or landunit-level behavior of anything else in CTSM, then I think this is a problem: in this case, I'm thinking that a vegetated landunit that has zero weight on the gridcell would operate fundamentally differently because col%wtgcell would be exactly 0 for all of its columns, so these other terms would be exactly 0. So if these terms feed back onto other behavior, then I think we'll need to change the way these are specified so that they are on a per-area basis (like most other variables in the model).

Reply:
Deferring this to https://github.com/ESCOMP/CTSM/issues/1805